### PR TITLE
Extract background images

### PR DESCRIPTION
**Pull request by kerrmarin**

### CODE REVIEW

File: lib/processLayerStyles.js [hunk 1]
```diff
@@ -72,12 +75,62 @@ const layerFill = (layerStyle) => {
 
         }
 
+        if (fill.fillType == 4 && fill.isEnabled == true) { // Images
```
Comment (deanWombourne) [inline lib/processLayerStyles.js:78]: Lol, love a good Magic Number :) What are the other fill types - do we need to add them to a list of future features to implement?

File: lib/processLayerStyles.js [hunk 2]
```diff
@@ -72,12 +75,62 @@ const layerFill = (layerStyle) => {
...
+          // Copy the image from the sketch file to the output folder
+          let imageOutputFolder = outputFolder + "/assets/Assets.xcassets"
+          let imageName = fill.image._ref.split("/").splice(-1)[0]
```
Comment (deanWombourne) [inline lib/processLayerStyles.js:81]: Can this be explained? I _think_ it's saying take the ref, break it down using "/", remove the last element and then take the first - but that doesn't make any sense?

File: lib/processLayerStyles.js [hunk 3]
```diff
@@ -72,12 +75,62 @@ const layerFill = (layerStyle) => {
...
 }
 
+const prepareAssetBundle = (imageName, assetBundleFolder) => {
+  const extension = imageName.split(".")[1]
+  const name = imageName.split(".")[0]
+  if (extension === "pdf") {
+    prepareAssetBundleForPDF(assetBundleFolder, name)
+  } else if (extension === "png") {
+    prepareAssetBundleForPNG(assetBundleFolder, name)
+  }
```
Comment (deanWombourne) [inline lib/processLayerStyles.js:131]: Should we have an `else { console.log("Unknown file extension " + extension) }` ?

File: lib/processLayerStyles.js (hunk 2 above)
Comment (kerrmarin) [reply lib/processLayerStyles.js:86]: Almost -- splice(-1) removes all items from the array except the last one, so you end up with [lastElement] and then take the first one which just so happens to be the only one (`[0]`)

Probably simpler/more undestandable as `fill.image._ref.split("/").pop()`?

File: lib/processLayerStyles.js (hunk 1 above)
Comment (kerrmarin) [reply lib/processLayerStyles.js]: @cdemetriadis do you know what the missing numbers are for fill types?

File: lib/processLayerStyles.js (hunk 2 above)
Comment (deanWombourne) [reply lib/processLayerStyles.js]: Ah, I'd assumed (and thought I'd tested but obviously hadn't correctly!) that it took off the last element, not returned only the last element! Shame there isn't a `last` method on arrays :)

FWIW I much much prefer `pop()`

File: lib/processLayerStyles.js [hunk 4]
```diff
@@ -1,17 +1,25 @@
-const { addOptionalProperty, colorToRGBA, doRound } = require('./processShared')
+const { addOptionalProperty, colorToRGBA, doRound, prepareAssetBundleForPDF, prepareAssetBundleForPNG } = require('./processShared')
+const fs = require("fs")
 
+const FILL_TYPE = {
+  SOLID_COLOR: 0,
+  GRADIENT: 1,
+  IMAGE: 4,
+}
```
Comment (deanWombourne) [inline lib/processLayerStyles.js:8]: ha ha this has bit mask written all over it - what happened to 2? Also, not sure why they've chosen a bit mask given you can't really combine them!

File: lib/processLayerStyles.js [hunk 5]
```diff
@@ -1,17 +1,25 @@
...
 
-const processLayerStyles = (layerStyles, filepath, outputFolder, modifiers) => {
+const processLayerStyles = async (layerStyles, filepath, outputFolder, modifiers) => {
     let ret = []
 
-    layerStyles.forEach((style)=>{
+    let sketchFile = await require('sketch-to-svg').SketchFile.from(filepath)
```
Comment (atomoil) [inline lib/processLayerStyles.js:13]: inline requires are a bit untidy - can it be moved to where the other requires are kept?

File: lib/processLayerStyles.js [hunk 6]
```diff
@@ -26,19 +34,19 @@ const processLayerStyles = (layerStyles, filepath, outputFolder, modifiers) => {
...
+const layerFill = (layerStyle, outputFolder, sketchFile) => {
     let fills = layerStyle.fills
 
     if (fills !== undefined && fills.length > 0) {
       let backgroundArray = []
       for(let i=0;i<fills.length;i++){
         let fill = fills[i]
 
-        if (fill.fillType == 0 && fill.isEnabled == true) { // Solid Color
+        if (fill.fillType == FILL_TYPE.SOLID_COLOR && fill.isEnabled == true) {
```
Comment (atomoil) [inline lib/processLayerStyles.js:45]: 💖

File: lib/processLayerStyles.js [hunk 7]
```diff
@@ -72,10 +80,61 @@ const layerFill = (layerStyle) => {
...
+            setTimeout(() => {
+              fs.copyFileSync(contentsPath + "/" + fill.image._ref, imageOutputFolder + `/${name}.imageset/` + imageName)
+            }, 1000)
+          })
+
+          // 0: do nothing
+          // 1: fill
+          // 2: stretch
+          // 3: fit
+          switch (fill.patternFillType) {
```
Comment (atomoil) [inline lib/processLayerStyles.js:105]: seeing as we have the nice FILL_TYPE enum could we have a PATTERN_FILL_TYPE enum too?

File: lib/processLayerStyles.js [hunk 8]
```diff
@@ -72,10 +80,61 @@ const layerFill = (layerStyle) => {
 
         }
 
+        if (fill.fillType == FILL_TYPE.IMAGE && fill.isEnabled == true) {
+          // Copy the image from the sketch file to the output folder
```
Comment (atomoil) [inline lib/processLayerStyles.js:84]: probably out of scope, but layerFill is now quite large - would it be possible to break it down into smaller functions for readability?

File: lib/processLayerStyles.js [hunk 9]
```diff
@@ -72,10 +80,61 @@ const layerFill = (layerStyle) => {
...
       }
       return backgroundArray
     }
-    // return "Color.clear"
+}
+
+const prepareAssetBundle = (imageName, assetBundleFolder) => {
+  const extension = imageName.split(".")[1]
+  const name = imageName.split(".")[0]
+  if (extension === "pdf") {
```
Comment (atomoil) [inline lib/processLayerStyles.js:131]: is it **always** lowercase, or do we need extension.toLowercase() on it?

File: lib/processShared.js [hunk 10]
```diff
@@ -32,4 +34,41 @@ function colorToRGBA(red, green, blue, alpha) {
...
+            { idiom: "universal", 
+            filename: `${name}${extension}` }
+        ],
+        info: { version:1, author:"stylableSwiftUI" }, 
+        properties: { "template-rendering-intent": "original" }
+    })
+    fs.writeFileSync(folder + "Contents.json", json)
+    return folder + name + extension
+    /*
+    pdfPath = path.resolve(pdfPath)
```
Comment (atomoil) [inline lib/processShared.js:60]: can this commented code be removed?

File: lib/processShared.js [hunk 11]
```diff
@@ -32,4 +34,41 @@ function colorToRGBA(red, green, blue, alpha) {
...
+const prepareAssetBundleForAsset = (assetBundleFolder, name, extension) => {
+    //let name = path.basename(filename, ".pdf")
+    let folder = `${assetBundleFolder}/${name}.imageset/`
+    fs.mkdirSync(folder, { recursive: true })
+    let json = JSON.stringify({
+        images: [
+            { idiom: "universal", 
+            filename: `${name}${extension}` }
+        ],
+        info: { version:1, author:"stylableSwiftUI" },
```
Comment (atomoil) [inline lib/processShared.js:54]: author: 'sketch-to-swiftui' ?

File: lib/processLayerStyles.js (hunk 9 above)
Comment (kerrmarin) [reply lib/processLayerStyles.js:131]: Not sure, but for safety I'll add the lowercase

File: lib/processLayerStyles.js (hunk 8 above)
Comment (kerrmarin) [reply lib/processLayerStyles.js:92]: I'll do that in another PR

**Review by atomoil — APPROVED** [review]

**Pull request merged**